packages/synapse-react-client/src/utils/hooks/useUploadFileEntity/useTrackFileUploads.ts

```diff
@@ -0,0 +1,182 @@
+    Map<File, TrackedUploadProgress>
+  >(new Map())
+
+  /**
+   * Adds new files to be tracked and puts them in the "PREPARED" state.
+   * Returns the new map of TrackedUploadProgress in case the new value is needed before
+   * the state variable is updated.
+   * @param preparedFiles
+   */
+  function trackNewFiles(
+    ...preparedFiles: FilePreparedForUpload[]
+  ): Map<File, TrackedUploadProgress> {
+    // If a matching file that is uploading already exists, cancel it!
+    preparedFiles.forEach(preparedFile => {
+      const existingUpload = trackedUploadProgress.get(preparedFile.file)
+      if (existingUpload && existingUpload.status !== 'CANCELED_BY_USER') {
+        cancelUpload(preparedFile.file)
+      }
+    })
+
+    // Re-initialize progress tracking state variable
+    const newTrackedUploadProgress: Map<File, TrackedUploadProgress> = new Map(
+      trackedUploadProgress.entries(),
+    )
+
+    preparedFiles.forEach(preparedFile => {
+      newTrackedUploadProgress.set(preparedFile.file, {
+        parentId: preparedFile.parentId,
+        abortController: new AbortController(),
+        // Note that this is number of parts uploaded, not file size or %
+        progress: { value: 0, total: 1 },
+        status: 'PREPARING',
+      })
+    })
+
+    setTrackedUploadProgress(newTrackedUploadProgress)
+
+    return newTrackedUploadProgress
+  }
+
+  function setProgress(file: File, progress: ProgressCallback) {
+    setTrackedUploadProgress(prev => {
+      const newMap = new Map(prev.entries())
+      const entry = newMap.get(file)
+      if (entry) {
+        newMap.set(file, {
+          ...entry,
+          progress: {
+            ...entry.progress,
+            ...progress,
+          },
+        })
+      }
+      return newMap
+    })
+  }
+
+  function setStatus(
+    file: File,
+    status: TrackedUploadProgress['status'],
+    failureReason?: string,
+  ) {
+    setTrackedUploadProgress(prev => {
+      const newMap = new Map(prev.entries())
+      const entry = newMap.get(file)
+      if (entry) {
+        newMap.set(file, {
+          ...entry,
+          status,
+          failureReason,
+        })
+      }
+      return newMap
+    })
+  }
+
+  function setIsUploading(file: File) {
+    setStatus(file, 'UPLOADING')
+  }
+
+  function setComplete(file: File) {
+    setStatus(file, 'COMPLETE')
+  }
+
+  function setFailed(file: File, failureReason: string) {
+    setStatus(file, 'FAILED', failureReason)
+  }
+
+  function cancelUpload(file: File) {
+    const entry = trackedUploadProgress.get(file)
+    if (entry != null) {
+      entry.abortController.abort()
+    }
+    setStatus(file, 'CANCELED_BY_USER')
+  }
+
+  function pauseUpload(file: File) {
+    const entry = trackedUploadProgress.get(file)
+    if (entry != null) {
+      entry.abortController.abort()
+    }
+    setStatus(file, 'PAUSED')
+  }
+
+  function removeUpload(file: File) {
```

Comment:
```suggestion
  /**
   * Removes the file from the list if the upload state is CANCELED_BY_USER or FAILED. Otherwise, a noop.
   */
  function removeUpload(file: File) {
```